xarray/core/missing.py

```diff
@@ -207,52 +207,79 @@ def _apply_over_vars_with_dim(func, self, dim=None, **kwargs):
+
+    Returns
+    -------
+    Variable
+      Numerical values for the x-coordinates.
 
-    If use_coordinate is False, the x values are set as an equally spaced
-    sequence.
+    Notes
+    -----
+    If indexing is along the time dimension, datetime coordinates are converted
+    to time deltas with respect to 1970-01-01.
     """
-    if use_coordinate:
-        if use_coordinate is True:
-            index = arr.get_index(dim)
-        else:
-            index = arr.coords[use_coordinate]
-            if index.ndim != 1:
-                raise ValueError(
-                    f"Coordinates used for interpolation must be 1D, "
-                    f"{use_coordinate} is {index.ndim}D."
-                )
-            index = index.to_index()
-
-        # TODO: index.name is None for multiindexes
-        # set name for nice error messages below
-        if isinstance(index, pd.MultiIndex):
-            index.name = dim
-
-        if not index.is_monotonic:
-            raise ValueError(f"Index {index.name!r} must be monotonically increasing")
-
-        if not index.is_unique:
-            raise ValueError(f"Index {index.name!r} has duplicate values")
-
-        # raise if index cannot be cast to a float (e.g. MultiIndex)
-        try:
-            index = index.values.astype(np.float64)
-        except (TypeError, ValueError):
-            # pandas raises a TypeError
-            # xarray/numpy raise a ValueError
-            raise TypeError(
-                f"Index {index.name!r} must be castable to float64 to support "
-                f"interpolation, got {type(index).__name__}."
-            )
 
-    else:
+    # Question: If use_coordinate is a string, what role does `dim` play?
+    from xarray.coding.cftimeindex import CFTimeIndex
+
+    if use_coordinate is False:
         axis = arr.get_axis_num(dim)
-        index = np.arange(arr.shape[axis], dtype=np.float64)
+        return np.arange(arr.shape[axis], dtype=np.float64)
+
+    if use_coordinate is True:
+        index = arr.get_index(dim)
+
+    else:  # string
+        index = arr.coords[use_coordinate]
+        if index.ndim != 1:
+            raise ValueError(
+                f"Coordinates used for interpolation must be 1D, "
+                f"{use_coordinate} is {index.ndim}D."
+            )
+        index = index.to_index()
+
+    # TODO: index.name is None for multiindexes
+    # set name for nice error messages below
+    if isinstance(index, pd.MultiIndex):
+        index.name = dim
+
+    if not index.is_monotonic:
+        raise ValueError(f"Index {index.name!r} must be monotonically increasing")
+
+    if not index.is_unique:
+        raise ValueError(f"Index {index.name!r} has duplicate values")
+
+    # Special case for non-standard calendar indexes
+    # Numerical datetime values are defined with respect to 1970-01-01T00:00:00 in units of nanoseconds
+    if isinstance(index, CFTimeIndex):
+        offset = type(index[0])(1970, 1, 1)
```

Comment:
how about `offset = index[0]` and `datetime_unit="D"` down below so we actually support a large time range?

cc @spencerkclark

Reply:
I think the rationale was to have a universal offset, so that fitted coefficients can be applied to other time series without having to specify the original offset explicitly. This is what happens for the standard pandas based calendar.    
```
coef = xr.polyfit(arr, deg=2)
out = xr.polyval(arr2.time, coef)
```

I tried to set `datetime_unit="D"` but still got an overflow error coming from pd.TimedeltaIndex.

Reply:
Yes, we need a universal offset for an application like `polyfit`/`polyval`.  For `interp` we just need to make sure we use a consistent offset for the reference coordinate and the coordinate we are interpolating to (both of which are known at the time the method is called, so it's fine to use the minimum value of the reference coordinate).

@huard the overflow error you're getting I think stems from passing the `CFTimeIndex` directly to `datetime_to_numeric`.  In that case subtracting the offset from the index will try to produce a `TimedeltaIndex` due to the logic here:
https://github.com/pydata/xarray/blob/471a5d6ce3eb10a521c8eebd3b97f72461405f19/xarray/coding/cftimeindex.py#L432-L433
You can work around this by passing the array of times instead of the index to `datetime_to_numeric`, e.g. using `index.values`<sup>1</sup>, though you will still run into the issue I raised in #3641.  I think we should try to address that first.

<sup>1 </sup>I also wonder to some extent whether we should fall back to just returning a generic index of `datetime.timedelta` objects in the case that we can't represent the differences using a `TimedeltaIndex`.  This might be friendlier than the current behavior, https://github.com/pydata/xarray/issues/3535, even though it might be a little trickier to work with.

Reply:
> we can't represent the differences using a TimedeltaIndex

At the very least we should raise a friendlier error message.